File: R/d07.r
box::use(
    dplyr[group_by, ungroup, summarise]
)

#' @export
solve_p1 = function(input) {
    dt = ._parse_input(input)
    return(sum(dt[dt$size < 100000,]$size))
}

#' @export
solve_p2 = function(input, req_space) {
    dt = ._parse_input(input)
    req_space = abs(40000000 - dt[dt$path == "/",]$size)
    return(min(dt[dt$size > req_space,]$size))
}

._parse_input = function(input) {
    paths = vector("list", length(input))
    size = vector("list", length(input))

    wd = NULL

    dt = data.frame(path=NULL,size=NULL) 

        for (ii in seq_len(length(input))) {
          x = input[ii]
          if (substr(x,1,7) == "$ cd ..") {
            wd = wd[1:(length(wd)-1)]
          } else if (substr(x,1,4) == "$ cd") {
            wd = c(wd, substr(x,6,nchar(x)))
          } else if (grepl("[[:digit:]]",substr(x,1,1))) {
            tmp = unlist(strsplit(x, " "))
            for (jj in 1:length(wd)) {
              dt = rbind(dt, data.frame(path=paste0(wd[1:jj],collapse = "/"),size=tmp[1]))
            }
          } else {
            next
          }
        }

        dt$size = as.integer(dt$size)
        dt = group_by(dt, path)
        dt = summarise(dt, size = sum(size))
        
        return(dt)
}

File: Python/d07.py
with open("data/d07.txt") as f:
    input = f.read().split("\n")

paths = []
size = []
wd = []

for ii in range(len(input)):
    x = input[ii]
    if x[0:7] == "$ cd ..":
        wd = wd[0:len(wd)-1]
    elif x[0:4] == "$ cd":
        wd.append(x[5:])
    elif x[0].isdigit():
        for l in range(len(wd)):
            paths.append("".join(wd[0:l+1]))
            size.append(x.split(" ")[0])            

uni_path = list(set(paths))
size_by_path = {}

for p in uni_path:
    total = 0
    for ii, val in enumerate(paths):
        if val == p:
            total += int(size[ii])
    size_by_path[p] = total

req_space = abs(40000000 - size_by_path["/"])

# Part 1 Answer
sum([s for s in size_by_path.values() if s <= 100000])

# Part 2 Answer
min([s for s in size_by_path.values() if s > req_space])
